docs/adr/2023-09-08_026-network-resilience.md

```diff
@@ -0,0 +1,68 @@
+  * There is a minor limitation, namely that the subscription mechanism does not handle connections invidually, but as a set of equivalent point-to-point full duplex connections whose size (valency) needs to be maintained at a certain threshold, which means that unless backed in the protocol itself, protocol state-machine and applications are not aware of the identity of the remote peer
+* We have built our `Network` infrastructure over the concept of relatively independent layers, each implementing a similar interface with different kind of messages, to `broadcast` messages to all peers and be notified of incoming messages through a `callback`.
+  * This pipes-like abstraction allows us to compose our network stack like:
+
+    ```
+     withAuthentication (contramap Authentication tracer) signingKey otherParties $
+      withHeartbeat nodeId connectionMessages $
+        withOuroborosNetwork (contramap Network tracer) localhost peers
+    ```
+
+  * This has the nice property that we can basically swap the lower layers should we need to, for example to use [UDP](https://github.com/input-output-hk/hydra/blob/abailly-iohk/multi-node-udp/hydra-node/src/Hydra/Network/UDP.hs), or add other layers for example to identify [Multiple Heads](https://github.com/input-output-hk/hydra/blob/abailly-iohk/multi-node-udp/hydra-node/src/Hydra/Network/MultiHead.hs#L26)
+
+## Decision
+
+* We implement our own message tracking and resending logic as a standalone `Network` layer
+* That layer consumes and produces `Authenticate`d messages in order to identify the source of messages
+* It uses a vector of monotonically increasing _sequence numbers_ associated with each party (including itself) to track what are the last messages from each party and to ensure FIFO delivery of messages
+* This _vector_ is also used to identify peers which are lagging behind and to resend the missing messages, or to drop messages which have already been received
+* This _vector_ is also piggybacked by the ping messages of the heartbeat mechanism so that a node informs its peers of its own situation even if it would have, otherwise, no message to sent to them
+* Sending a ping message does not influence the _vector_ of the sender
+* Any message received which index does not match what the peer expects is dropped
+* Messages deemed not received by a peer a re-sent
+
+## Consequences
+
+* We keep our existing `Network` interface hence all messages will be resent to all peers
+  * This could be later optimized either by providing a smarter interface with a `send :: Peer -> msg -> m ()` unicast function, or by adding a layer with filtering capabilities, or both
+* We need to ensure messages are not kept forever when resending, eg. that the pending messages list is garbage collected
```

Comment:
Do we want to specify this further? i.e. To me, this consequence implies a *configurable* parameter to the `Resending` network layer, limiting how many messages are kept as a maximum.

Reply:
TL;DR in my opinion, this should be considered in the broader context of automatic head closing conditions for a node.

I wouldn't specify it here. In current implement we do not have any limitation on how many messages can be kept as a maximum. The reason is that we can garbage collect these messages when we're sure they've been received.

So the number of messages stored, in practice, is not limited and is dependent of the hydra network _quality_. If all peers are behaving correctly, on a working network, this number is exeptec to be very low. If, for some reason, the number of messages kept for re-sending them grows, it means we have, either a network problem, or a node being down. Then we should really keep all the messages until communication is restored or the head will get stuck.

So the question, I think, becomes rather how bad do we tolerate our peer network to be before closing the head, rather than dropping messages.

Reply:
I have removed the item